### PR TITLE
Support psycopg3

### PR DESCRIPTION
**Pull request by snuderl**
Very minor changes needed to support this.

I am not sure what the best way to add tests for this would be, as ideally test would ran against both psycopg2 and psycopg3.

Edit:
Set up the tests.
- django 4.2 does not support postgres 11 anymore so I bumped that
- if both psycopg and psycopg2 are installed, django will pick psycopg so I just install it alongside. Though this does bring up a point. Presumably we should not require psycopg2 anymore if psycopg is installed, but I am not entirely sure how to express this.
- ran tests here https://app.circleci.com/pipelines/github/snuderl/django-postgres-extra/13/workflows/fd538c23-1990-42c5-b6d8-889b54262625 Everything is green except the coveralls hook

P.S.
Really sweet testing setup.

### CODE REVIEW

File: psqlextra/compiler.py [hunk 1]
```diff
@@ -187,12 +187,13 @@ def execute_sql(self, return_id=False):
...
                 except ProgrammingError:
                     pass
+            description = cursor.description
 
         # create a mapping between column names and column value
         return [
             {
                 column.name: row[column_index]
-                for column_index, column in enumerate(cursor.description)
+                for column_index, column in enumerate(description)
```
Comment (Photonios) [inline psqlextra/compiler.py:196]: I don't understand this change. How is aliasing it on line 190 changing how this works?

**Comment by Photonios** [comment]
> I am not sure what the best way to add tests for this would be, as ideally test would ran against both psycopg2 and psycopg3.

This could be fixed by adding a new dimension to the Tox configuration file which already runs the tests against multiple Django and Python versions. See: https://github.com/SectorLabs/django-postgres-extra/blob/master/tox.ini

**Review by Photonios — COMMENTED** [review]
👍

File: psqlextra/compiler.py (hunk 1 above)
Comment (snuderl) [reply psqlextra/compiler.py:196]: It seems that the `with cursor` now cleans up the data on the cursor on exit. So `cusror.description` ends up being None. So we just need to read it before the with block exits

File: .circleci/config.yml [hunk 2]
```diff
@@ -78,7 +84,7 @@ jobs:
           extra: test
       - run-tests:
           pyversion: 38
-          djversions: 20,21,22,30,31,32,40
+          djversions: 20,21,22,30,31,32,40,42
```
Comment (damjankuznar) [inline .circleci/config.yml:87]: why not also 41?

**Review by damjankuznar — DISMISSED** [review]
lgtm

File: .circleci/config.yml [hunk 3]
```diff
@@ -42,6 +42,12 @@ commands:
           environment:
             DATABASE_URL: 'postgres://psqlextra:psqlextra@localhost:5432/psqlextra'
 
+      - run:
+          name: Run tests psycopg3
+          command: tox -e 'py<< parameters.pyversion >>-dj{<< parameters.djversions >>}'
```
Comment (Photonios) [inline .circleci/config.yml:47]: How does this install psycopg3? I don't see any references in `tox.ini` to psycopg3.

Comment (snuderl) [reply .circleci/config.yml:47]: oh sorry, this was me trying to figure tox/circle out

Right now the djversion=42 uses psycopg3 (though It means that it does not test psycopg2)

**Review by Photonios — APPROVED** [review]

File: tox.ini [hunk 4]
```diff
@@ -11,6 +11,7 @@ deps =
     dj32: Django~=3.2.0
     dj40: Django~=4.0.0
     dj41: Django~=4.1.0
+    dj42: .[psycopg3]
```
Comment (Photonios) [inline tox.ini:14]: This means the tests don't run against psycopg2 anymore. I will patch this up after merging this PR.

**Review by Photonios — APPROVED** [review]

File: tox.ini (hunk 4 above)
Comment (Photonios) [reply tox.ini:14]: https://github.com/SectorLabs/django-postgres-extra/pull/209

I just added this. Tests run against psycopg 2.8, 2.9 and 3.0 now with all supported Django versions.

Comment (snuderl) [reply tox.ini:14]: Thanks!

**Comment by Photonios** [comment]
@snuderl You can try it out in `v2.0.9rc2`

**Comment by snuderl** [comment]
> @snuderl You can try it out in `v2.0.9rc2`

Yup. Works with our codebase!

**Comment by adamantike** [comment]
I was also experiencing this issue, and bumping to one of the current RC releases fixed it. Looking for a new release to unpin our dependency from RC! :slightly_smiling_face:

**Pull request merged**